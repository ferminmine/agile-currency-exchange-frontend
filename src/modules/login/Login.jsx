import React from 'react';
import { connect } from 'react-redux';
import { compose, bindActionCreators } from 'redux';
import withStyles from 'react-jss';
import { push } from 'connected-react-router';
import Swal from 'sweetalert2';
import styles from './LoginStyles';
import { attemptUserLogin } from '../../utils/ApiClient';
import { fetchUserInfo } from '../user/UserActions';
import PropTypes from 'prop-types';

class Login extends React.Component {
  static propTypes = {
    push: PropTypes.func
  };

  state = {
    username: '',
    password: ''
  };

  onInputChange = event => {
    this.setState({ [event.target.name]: event.target.value });
  };

  sweetAlertLoginMessage = () => ({
    showCancelButton: false,
    confirmButtonText: 'Continue',
    allowOutsideClick: false
  })

  attemptLogin = async () => {
    const { username, password } = this.state
    const token = await attemptUserLogin(username, password);
    const sweetAlertLoginMessage = this.sweetAlertLoginMessage();
    if (token) {
      await Swal.fire({
        ...sweetAlertLoginMessage,
        title: 'Welcome back!',
        text: 'You just successfully logged in. We are happy to see you again!',
        type: 'success'
      });
      this.props.fetchUserInfo(JWTDecode(token).user_id);
      this.props.push('/');
    } else {
      await Swal.fire({
        ...sweetAlertLoginMessage,
        title: 'Error trying to Log In',
        text: 'There has been an error while trying to log in. Please check your credentials or try again later.',
        type: 'error'
      });
    }
  }

  render = () => {
    const { classes } = this.props;
    return (
      <div className={classes.loginContainer}>
        <div className={classes.formContainer}>
          <h2 className={classes.loginTitle}> Login </h2>
          <input
            className={classes.loginInput}
            type="text"
            placeholder="Username"
            name="username"
            onChange={this.onInputChange}
          ></input>
          <input
            className={classes.loginInput}
            type="password"
            placeholder="Password"
            name="password"
            onChange={this.onInputChange}
          ></input>
        </div>
        <button className={classes.submitButton} onClick={this.attemptLogin}> Submit </button>
      </div>
    );
  };
}

const JWTDecode = token => {
  try {
    return JSON.parse(atob(token.split('.')[1]));
  } catch (e) {
    return null;
  }
};

const mapDispatchToProps = dispatch => bindActionCreators({
  push,
  fetchUserInfo
}, dispatch)


export default compose(
  withStyles(styles),
  connect(null, mapDispatchToProps)
)(Login);
