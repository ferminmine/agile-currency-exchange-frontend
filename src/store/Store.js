import { createStore, combineReducers, compose, applyMiddleware } from 'redux';
import { createBrowserHistory } from 'history';
import thunk from 'redux-thunk';
import { connectRouter, routerMiddleware } from 'connected-react-router';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import hardSet from 'redux-persist/lib/stateReconciler/hardSet';
import userReducer from '../modules/user/UserReducer';
import currenciesReducer from '../modules/currencies/CurrenciesReducer';
import accountReducer from '../modules/account/AccountReducer';

export const history = createBrowserHistory();

const persistConfig = {
  key: 'root',
  storage,
  blacklist: ['router'],
  stateReconciler: hardSet
};

const allReducers = combineReducers({
  user: userReducer,
  currencies: currenciesReducer,
  account: accountReducer,
  router: connectRouter(history)
});

export const ROOT_REDUCER_ACTION_TYPES = {
  RESET_APP: 'RESET_APP'
};
export const resetApp = () => ({ type: ROOT_REDUCER_ACTION_TYPES.RESET_APP });

const rootReducer = (state, action) => {
  if (action.type === 'RESET_APP') {
    state = undefined;
  }

  return allReducers(state, action);
};

const persistedReducer = persistReducer(persistConfig, rootReducer);
const composeEnhancer = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const store = createStore(
  persistedReducer,
  composeEnhancer(applyMiddleware(routerMiddleware(history), thunk))
);
export default store;
export const persistor = persistStore(store);
