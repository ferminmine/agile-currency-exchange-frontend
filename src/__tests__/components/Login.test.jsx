import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import Login from '../../modules/login/Login';
import store from '../../store/Store';

describe('Login Component Test', () => {
  it('renders without crashing', () => {
    mount(
      <Provider store={store}>
        <Login />
      </Provider>
    );
  });

  it('matches the snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Login />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
