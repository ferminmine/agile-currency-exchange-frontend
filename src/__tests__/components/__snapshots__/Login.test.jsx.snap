// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component Test matches the snapshot 1`] = `
<div
  className="Connect(Login)-loginContainer-0-1-4"
>
  <div
    className="Connect(Login)-formContainer-0-1-2"
  >
    <h2
      className="Connect(Login)-loginTitle-0-1-5"
    >
       Login 
    </h2>
    <input
      className="Connect(Login)-formInput-0-1-3"
      name="username"
      onChange={[Function]}
      placeholder="Username"
      type="text"
    />
    <input
      className="Connect(Login)-formInput-0-1-3"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
    />
  </div>
  <button
    className="Connect(Login)-submitButton-0-1-1"
    onClick={[Function]}
  >
     Submit 
  </button>
</div>
`;
